### PR TITLE
[ArmSME][test] Enable TransposeMatmulPass and peeling for e2e matmuls

### PR DESCRIPTION
**Pull request by MacDue**
ci-extra: build_test_all_arm64

### CODE REVIEW

File: tests/e2e/matmul/CMakeLists.txt
Comment (ScottTodd) [inline tests/e2e/matmul/CMakeLists.txt:61]: What are the eventual goals for `TransposeMatmulPass`? I'm concerned about these sorts of verbose "preprocessing pass pipelines" making their way into test configurations and scripts unconstrained. We shouldn't require users (e.g. from the Python bindings) to spell out these explicit compiler passes.

Is this on a path to having its own top level option? Is it something that can be enabled unconditionally once more testing is done? Is it something that the compiler can decide when to enable on its own?

Comment (MacDue) [reply tests/e2e/matmul/CMakeLists.txt:61]: I'm not entirely sure... This was initially added as a global optimization (but that was rejected). It's something that should always be enabled when lowering a matmul to ArmSME, though IIRC the target information is not available early on in the pipeline (so it can't be automatically enabled). @c-rhodes and @MaheshRavishankar may have more context here.

Comment (MaheshRavishankar) [reply tests/e2e/matmul/CMakeLists.txt:61]: The transposition done here is a short cut. Ideally this should all just fold into data-tiling. This is not something we want to enable by default in the compiler. This is kind of an opt-in transformation at the global level for a specific target hardware.
I would really think it would be better if this was a test downstream. Maybe a good structuring for the ARM/SVE/SME work is to use a plugin where you can have such opinionated transformations on by default and tested in the plugin like we do for the iree-amd-aie plugin.

Having said that, this test is only being run on ARM, so I am fine with it, but I think Scott has the final say here.

Comment (c-rhodes) [reply tests/e2e/matmul/CMakeLists.txt:61]: >The transposition done here is a short cut. Ideally this should all just fold into data-tiling. This is not something we want to enable by default in the compiler. This is kind of an opt-in transformation at the global level for a specific target hardware.

+1 the preprocessing pass gives us a big win for minimal effort. As mentioned in [1] we'll be looking at alternatives such as data-tiling going forward.

>I would really think it would be better if this was a test downstream. Maybe a good structuring for the ARM/SVE/SME work is to use a plugin where you can have such opinionated transformations on by default and tested in the plugin like we do for the iree-amd-aie plugin.

We have no downstream.

[1] https://github.com/iree-org/iree/pull/17157#discussion_r1580541805

Comment (ScottTodd) [reply tests/e2e/matmul/CMakeLists.txt:61]: Okay, sounds like you all have a handle on the tradeoffs involved (ease of integration via preprocessing, longer term alternatives to graduate from there). This LGTM then, the new test coverage is worth the extra bit of infra complexity here.

**Review by ScottTodd — APPROVED** [review]

**Comment by MacDue** [comment]
Could someone land this for me? I don't have write access. 
Thanks :pray:

**Pull request merged**
